Add expireTag and expirePath APIs (#72485)
vercel/next.js: docs/01-app/03-api-reference/04-functions/expirePath.mdx

---
title: expirePath

> description: API Reference for the expirePath function.

Reviewer: Might want to [mark these as canary](https://vercel.slack.com/archives/C06DGJ420TZ/p1730990131156869?thread_ts=1730208016.298569&cid=C06DGJ420TZ) until the next minor release? 

```suggestion
description: API Reference for the expirePath function.
version: canary
```

Author: Won't these only show on the canary docs version until we cut a stable release?

---

`expirePath` allows you to purge [cached data](/docs/app/building-your-application/caching) on-demand for a specific path.

> **Good to know**:
>
> - `expirePath` is available in both [Node.js and Edge runtimes](/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes).

> > - `expirePath` only invalidates the cache when the included path is next visited. This means calling `expirePath` with a dynamic route segment will not immediately trigger many revalidations at once. The invalidation only happens when the path is next visited.

Reviewer 2: ```suggestion
> - `expirePath` only invalidates the cache when the included path is next visited. This means calling `expirePath` with a dynamic route segment will not immediately trigger many expirations at once. The invalidation only happens when the path is next visited.
```

> > - Currently, `expirePath` invalidates all the routes in the [client-side Router Cache](/docs/app/building-your-application/caching#client-side-router-cache). This behavior is temporary and will be updated in the future to apply only to the specific path.

Author: Yea this is from existing doc can update for both though

> > - Currently, `expirePath` invalidates all the routes in the [client-side Router Cache](/docs/app/building-your-application/caching#client-side-router-cache). This behavior is temporary and will be updated in the future to apply only to the specific path.

Author: updated note in https://github.com/vercel/next.js/pull/72485/commits/39d213a997431536905259aa71a9889289e1ce51

> > - Currently, `expirePath` invalidates all the routes in the [client-side Router Cache](/docs/app/building-your-application/caching#client-side-router-cache) when used in a server action. This behavior is temporary and will be updated in the future to apply only to the specific path.

Reviewer 3: Q: Do we have a warning for that? Sounds like easily being misused if the path is set, but all client cache is invalidated?

Author: That would be a warning for any time `expirePath()` or `revalidatePath()` is used in a server action so too noisy.

> - Using `expirePath` invalidates **only the specific path** in the [server-side Route Cache](/docs/app/building-your-application/caching#full-route-cache).

## Parameters

```tsx
expirePath(path: string, type?: 'page' | 'layout'): void;
```

> - `path`: Either a string representing the filesystem path associated with the data you want to revalidate (for example, `/product/[slug]/page`), or the literal route segment (for example, `/product/123`). Must be less than 1024 characters. This value is case-sensitive.

Reviewer 2: ```suggestion
- `path`: Either a string representing the filesystem path associated with the data you want to expire (for example, `/product/[slug]/page`), or the literal route segment (for example, `/product/123`). Must be less than 1024 characters. This value is case-sensitive.
```

> - `type`: (optional) `'page'` or `'layout'` string to change the type of path to revalidate. If `path` contains a dynamic segment (for example, `/product/[slug]/page`), this parameter is required. If path refers to the literal route segment, e.g., `/product/1` for a dynamic page (e.g., `/product/[slug]/page`), you should not provide `type`.

Reviewer 2: ```suggestion
- `type`: (optional) `'page'` or `'layout'` string to change the type of path to expire. If `path` contains a dynamic segment (for example, `/product/[slug]/page`), this parameter is required. If path refers to the literal route segment, e.g., `/product/1` for a dynamic page (e.g., `/product/[slug]/page`), you should not provide `type`.
```

## Returns

`expirePath` does not return a value.

## Examples

> ### Revalidating A Specific URL

Reviewer 2: ```suggestion
### Expiring A Specific URL
```

```ts
import { expirePath } from 'next/cache'
expirePath('/blog/post-1')
```

> This will revalidate one specific URL on the next page visit.

Reviewer 2: ```suggestion
This will purge the cache for one specific URL on the next page visit.
```

> ### Revalidating A Page Path

Reviewer 2: ```suggestion
### Expiring A Page Path
```

```ts
import { expirePath } from 'next/cache'
expirePath('/blog/[slug]', 'page')
// or with route groups
expirePath('/(main)/blog/[slug]', 'page')
```

> This will revalidate any URL that matches the provided `page` file on the next page visit. This will _not_ invalidate pages beneath the specific page. For example, `/blog/[slug]` won't invalidate `/blog/[slug]/[author]`.

Reviewer 2: ```suggestion
This will purge the cache any URL that matches the provided `page` file on the next page visit. This will _not_ invalidate pages beneath the specific page. For example, `/blog/[slug]` won't invalidate `/blog/[slug]/[author]`.
```

> ### Revalidating A Layout Path

Reviewer 2: ```suggestion
### Expiring A Layout Path
```

```ts
import { expirePath } from 'next/cache'
expirePath('/blog/[slug]', 'layout')
// or with route groups
expirePath('/(main)/post/[slug]', 'layout')
```

> This will revalidate any URL that matches the provided `layout` file on the next page visit. This will cause pages beneath with the same layout to revalidate on the next visit. For example, in the above case, `/blog/[slug]/[another]` would also revalidate on the next visit.

Reviewer 2: ```suggestion
This will purge the cache on any URL that matches the provided `layout` file on the next page visit. This will cause pages beneath with the same layout to revalidate on the next visit. For example, in the above case, `/blog/[slug]/[another]` would also revalidate on the next visit.
```

> ### Revalidating All Data

Reviewer 2: ```suggestion
### Expiring All Data
```

```ts
import { expirePath } from 'next/cache'

expirePath('/', 'layout')

> ```

Author: Yeah same as existing `revalidatePath`

> This will purge the Client-side Router Cache, and revalidate the Data Cache on the next page visit.

Reviewer 2: ```suggestion
This will purge the Data Cache on the next page visit.
```

### Server Action

```ts filename="app/actions.ts" switcher
'use server'

import { expirePath } from 'next/cache'

export default async function submit() {
  await submitForm()
  expirePath('/')
}
```

### Route Handler

> ```ts filename="app/api/revalidate/route.ts" switcher

Reviewer 2: ```suggestion
```ts filename="app/api/expire/route.ts" switcher
```

import { expirePath } from 'next/cache'
import type { NextRequest } from 'next/server'

export async function GET(request: NextRequest) {
  const path = request.nextUrl.searchParams.get('path')

  if (path) {
    expirePath(path)
    return Response.json({ revalidated: true, now: Date.now() })
  }

  return Response.json({

>     revalidated: false,

Reviewer 2: ```suggestion
    expired: false,
```

    now: Date.now(),

>     message: 'Missing path to revalidate',

Reviewer 2: ```suggestion
    message: 'Missing path to expire',
```

  })
}
```

> ```js filename="app/api/revalidate/route.js" switcher

Reviewer 2: ```suggestion
```js filename="app/api/expire/route.js" switcher
```

import { expirePath } from 'next/cache'

export async function GET(request) {
  const path = request.nextUrl.searchParams.get('path')

  if (path) {
    expirePath(path)

>     return Response.json({ revalidated: true, now: Date.now() })

Reviewer 2: ```suggestion
    return Response.json({ expired: true, now: Date.now() })
```

  }

  return Response.json({

>     revalidated: false,

Reviewer 2: ```suggestion
    expired: false,
```

    now: Date.now(),

>     message: 'Missing path to revalidate',

Reviewer 2: ```suggestion
    message: 'Missing path to expire',
```
